app/js/lib/intent-parser.js

```diff
@@ -0,0 +1,69 @@
+'use strict';
+
+/*
+Examples of phrases to detect and parse:
+[Remind] me [to] pick Sasha from Santa Clara University [at] 5PM today.
+[Remind] me [that] it is picnic day [on] July 4th.
+[Remind] us [to] go to the opera [at] 7:15pm on 2nd February.
+[Remind] us [to] go at my mum's [at] 11:30am on 31st July.
+[Remind] everybody [to] pack their stuff [by] next Friday 5pm.
+*/
+
+const PUNCTUATION_REGEXP = new RegExp('[.!?]+$', 'u');
+const pattern = {
+  regexp: new RegExp('^Remind\\b([^to]+)\\bto\\b(.+)\\bat\\b(.+)$', 'iu'),
+  placeholders: {
+    user: 0,
+    action: 1,
+    time: 2,
+  },
+};
+
+export default class IntentParser {
+  parse(string = '') {
+    string = this.normalise(string);
+
+    if (!string) {
+      return Promise.reject('Empty string.');
+    }
+
+    if (!pattern.regexp.test(string)) {
+      return Promise.reject('Unsupported intent format.');
+    }
+
+    const segments = pattern.regexp.exec(string);
+    segments.shift();
+
+    const user = this.extractRecipient(segments[pattern.placeholders.user]);
+    const action = this.extractAction(segments[pattern.placeholders.action]);
+    const time = this.extractDatetime(segments[pattern.placeholders.time]);
+
+    return Promise.resolve({ user, action, time });
+  }
+
+  extractRecipient(string = '') {
+    string = this.normalise(string);
```

Comment:
I don't think we should normalise the string everytime (it invokes to REs right now - could end up with more) - maybe just expect a normalised string?